### ISSUE TITLE
GitHub Application support

### ISSUE THREAD

**Issue by bradrydzewski** [harness/harness#2422]
For a while now I have wanted to enable support for [GitHub Applications](https://developer.github.com/apps/differences-between-apps/#about-github-apps) as an alternative to the current oauth application implementation. The introduction of GitHub Checks, which is only available to GitHub Applications, is increasing the priority of this feature.

Support for GitHub applications will require some changes to how Drone works. Here are some key differences between GitHub applications and oauth applications:

> GitHub Apps can look at /installation/repositories to see repositories the installation can access.

> GitHub Apps receive webhooks when repositories are added or removed from the installation.	

> By default, GitHub Apps have a single webhook that receives the events they are configured to receive for every repository they have access to.

> GitHub Apps ask for repository contents permission and use your installation token to authenticate via HTTP-based Git.

The 0.9 codebase implements a number of meaningful abstractions that should help simplify the transition to GitHub applications.

__Security Considerations__

The primary concern I have at this time is with regards to git access and cloning:

> GitHub Apps ask for repository contents permission and use your installation token to authenticate via HTTP-based Git. The token is used as the HTTP password.

The challenge here is the clone happens inside the vm. This means we need to send the token down to every agent and inject into the pipeline environment. This token grants access to __every__ repository and registered drone user. If an agent is compromised or the token is exposed, an attacker will potentially have access to significantly more data than they would with our current implementation.

__Token Expiration__

I should note that the tokens have a 1 hour expiration, however, we can assume that even a few minutes would be enough time for an attacker to clone every repository in the system with the stolen token.

This also brings me to the second design challenge. If the token expires every 1 hour, we need understand the impact of a token expiring midway through build execution. In a multi-stage build we can expect the pipeline to clone a repository or dependency using the token / netrc file many minutes into the build, and need to account for the fact that the token could be expired.

This has me wondering if we need to adjust the current behavior of Drone to use SSH keys to clone the repository, and provide repository owners an option to configure a user token that can be optionally used to clone additional repositories or dependencies. This token would optionally be exposed to the pipeline environment in a netrc file. This makes the exposure of the token opt-in, and prevents us from exposing the GitHub installation token to our agents, minimizing the impact of token theft.

cc @gtaylor @tboerger @metalmatze @tonglil any thoughts?

**Comment by andrewmclagan** [comment]
The complexity of a token expiring mid build for one hour token expirations is a pretty large edge case?

Although I'm assuming that someone.. somewhere will be running builds like this?

**Comment by tboerger** [comment]
Supporting apps seems to be important to get it properly integrated into github.

For the security perspective I would tend to ssh key or token per repository or maybe even on the organization level?

**Comment by tonglil** [comment]
SSH keys essentially changes the access model from an "all access" user personal-access-token (PAT) to a repo-specific key.

This means private git submodules and other private git repo cloning (npm/go/etc libraries) now will not work without additional authorization or authorized mirror (increased security, prevents laziness).

Could be a bit cumbersome to make modifications in places, but it does ensure that a token in one place does not grant wide access to other resources.

I was hoping that GitHub Apps would have an authorization method to allow each repo to use a dedicated token for that repo. Users can still upload their own PAT to Drone and create a `.netrc` file themselves.

**Comment by bradrydzewski** [comment]
Unfortunately github apps is worse from a security perspective for this particular use case, since you have to send the global application token down to the agent in order to clone the repository. I will probably engage github support before I finalize a decision, just to ensure I am not misunderstanding how this all works.

> Could be a bit cumbersome to make modifications in places

I was thinking we could add a dedicated textarea field in the user interface for providing a netrc file. This would add an extra step to the configuration if you clone private dependencies, but perhaps we can simplify the data entry and the runtime netrc file generation. We could also include the ability to provide a global netrc file, similar to global secrets and global registry credentials, to reduce this burden.

But yea, this is a classic case of security vs convenience. Unfortunately I am not sure there is a great way to provide both, but I am most certainly open to suggestions and very much hope to be proven wrong :)

**Comment by PatWie** [comment]
Isn't the token for each individual installation per repository? There is an installation Id for each registered repository and a single integration id.

For a very [basic setup](https://github.com/PatWie/pylint/blob/a0283500c424109de6c0614a2c28c3e45284483c/server.go#L112) I needed to store the per installation id to generate an access token.

**Comment by renothing** [comment]
since there is api for deploy key, why would we use deploy key for clone? it's readonly access permission.

**Comment by gtaylor** [comment]
Also forgot to ask: Are we using a global webhook instead of per-repo hooks?

**Comment by bradrydzewski** [comment]
@gtaylor yes, planning to use global webhooks. I am hoping this will be supported in the initial implementation, but we could fallback to per-repository webhooks to push out an mvp sooner. TBD

**Comment by krlmlr** [comment]
Thanks for implementing and providing this great CI solution. Is there an ETA for this particular feature?

**Comment by leosunmo** [comment]
Are there any tasks you could break down for us that need to be implemented before Github App support can continue? I'd be keen to help out with this as it's something my organisation really wants, but it's hard to get started on a new codebase with something this complex (judging from the previous comments it'll be a multi-step process to get this over the finish-line).

Any first steps that we could tackle would be much appreciated!

**Comment by mzaglia** [comment]
Any update on this issue? Looking forward for this

**Comment by Morriz** [comment]
Yeah, we are now stuck with an organization not allowed to give us an oauth app. A GitHub app with limited repo access would be the solution.

**Comment by zoey-juan** [comment]
Looking for the update as well

**Comment by tonglil** [comment]
Revisiting this, it seems like some kind of GH App support was added to the `scm` driver [here](https://github.com/drone/go-scm/pull/213/files), and the [code is integrated](https://github.com/search?q=%28org%3Adrone+OR+org%3Aharness%29+AND+ListByInstallation&type=code) with Harness's version of Drone [here](https://github.com/harness/harness-core/blob/01fdb869d92a40619e305e05dd13ddcd6d043bba/product/ci/scm/git/git.go#L564).

I wonder if there's a way to surface this use here so that non-Harness Drone installations can use it in places like https://github.com/harness/drone/blob/fdf51e70b2b592fbb4d38b31e5ab8add8455fe18/handler/api/user/sync.go#L56?

Or alternatively, could a different `scm` driver be created for GHAs..
